File: Leetcode/0530 Minimum Absolute Difference in BST.py
class Solution:
    # O(n) time, O(h) space
    # n = #nodes, h = height of the tree
    def getMinimumDifference(self, root):
        self.min_diff = float('inf')
        self.dfs(root, None)
        return self.min_diff

    def dfs(self, root, prev):
        if root is None:
            return prev
        
        prev = self.dfs(root.left, prev)
        
        if prev is not None:
            self.min_diff = min(self.min_diff, root.val - prev.val)

        return self.dfs(root.right, root)

File: Leetcode/0530 Minimum Absolute Difference in BST.ts
function getMinimumDifference(root: TreeNode | null): number {
    let prevNode: TreeNode | null = null, minDiff = Infinity;
    function dfs(root: TreeNode | null) {
        if (root === null) {
            return;
        }
        dfs(root.left);
        if (prevNode != null) {
            minDiff = Math.min(minDiff, root.val - prevNode.val);
        }
        prevNode = root;
        dfs(root.right);
    }
    dfs(root);
    return minDiff;
}
